Parser support for assigns contracts on loops

aalok-thakkar opened the pull request:
1. Created cprover_loop_contract that has cprover_contract_assigns_opt and loop_invariant_opt.

This adds support for parsing assigns clauses for loops.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

SaswatPadhi commented on src/ansi-c/parser.y:
```diff
@@ -3271,6 +3271,12 @@ cprover_function_contract_sequence_opt:
+cprover_loop_contract:
+          cprover_contract_assigns_opt
+          loop_invariant_opt
+        | loop_invariant_opt
```
Why do we need these two rules? Wouldn't the first rule suffice? `cprover_contract_assigns_opt` can expand to epsilon.

aalok-thakkar replied:
I have now created cprover_contract_assigns and cprover_contract_assigns_opt. This resolves the issue.

SaswatPadhi commented on src/ansi-c/parser.y:
```diff
@@ -3271,6 +3271,12 @@ cprover_function_contract_sequence_opt:
 
+cprover_loop_contract:
+          cprover_contract_assigns_opt
+          loop_invariant_opt
```
Could you please rename `loop_invariant_opt` to `cprover_contract_invariant_opt` for consistency? 

And, similarly, `cprover_contract_decreases_opt` when you resolve merge conflicts.

In the future, I would update the invariant keyword to `__CPROVER_invariant`, and we can specify both decreases and invariants in contracts for (recursive) functions as well.

aalok-thakkar replied:
Renamed, and merge conflicts resolved.

SaswatPadhi reviewed (changes requested):
Please rebase against current `develop` and resolve the merge conflict -- it's due to the decreases contract that Long's PR introduced. That should be added to `cprover_loop_contract` and should be removed from individual iteration statements.

feliperodri reviewed (approved):
LGTM.

SaswatPadhi commented:
Marking this as draft. There are several issues with commits, including merge conflicts.
@aalok-thakkar please fix these.

martin-cs reviewed (commented):
Please rebase as there seems to be at least one unrelated commit in this PR.

SaswatPadhi commented on src/ansi-c/parser.y:
```diff
@@ -3552,4 +3564,4 @@ parameter_postfix_abstract_declarator:
         ;
 
-%%
+%%
```
Please add this newline back.

aalok-thakkar replied:
Added.

SaswatPadhi reviewed (changes requested):
LGTM, but one minor issue.

martin-cs reviewed (changes requested):
The patch looks OK but I really don't think we should be adding syntax extensions without test cases that at least demonstrate how to use them.

feliperodri commented on regression/contracts/invar_loop_constant_pass/test.desc:
```diff
@@ -10,4 +10,4 @@ main.c
 --
 This test checks that the invariant is correctly used to satisfy
-all assertions that follow a loop.
+all assertions that follow a loop.
```
Fix end of line.

feliperodri replied:
?

feliperodri commented on regression/contracts/invar_dynamic_struct_member/main.c:
```diff
@@ -9,7 +9,7 @@ void main()
 
   unsigned n;
-  for(unsigned i = 0; i < n; ++i)
+  for(unsigned i = 0; i < n; ++i)  
```
Remove white space.

feliperodri replied:
?

feliperodri commented on regression/contracts/invar_check_multiple_loops/main.c:
```diff
@@ -26,4 +27,4 @@ int main()
 
   return 0;
-}
+}
```
Fix end of line.

feliperodri replied:
?

feliperodri commented on regression/contracts/invar_assigns_opt/test.desc:
```diff
@@ -0,0 +1,20 @@
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test checks that adding assigns clause is optional
```
Could you rebase this branch and add a test case using an empty assigns clause (e.g., `__CPROVER_assigns()`)?

feliperodri commented on regression/contracts/invar_assigns_opt/main.c:
```diff
@@ -0,0 +1,36 @@
+  assert(s2 == 1);
+
+  return 0;
+
```
Remove unnecessary empty spaces.

feliperodri commented on regression/contracts/invar_assigns_alias_analysis/test.desc:
```diff
@@ -0,0 +1,13 @@
+^VERIFICATION FAILED$
+--
+--
+This test shows a known bug in the current implementation
+for assigns clause. Fixing this is a part of future work.
```
Describe the bug.

Commit:
Parser support for assigns contracts on loops

1. Renamed cprover_contract_assigns_opt to cprover_contract_assigns.
2. Created cprover_contract_assigns_opt which has a unit production and cprover_contract_assigns.
3. Renamed cprover_decreases_opt to cprover_contract_decreases_opt.
4. Renamed loop_invariant to cprover_contract_loop_invariant.
5. Renamed loop_invariant_list to cprover_contract_loop_invariant_list.
6. Renamed loop_invariant_list_opt to cprover_contract_loop_invariant_list_opt.
7. Added cprover_contract_assigns_opt to iteration statement.
8. Created the regression test invar_assigns_alias_analysis (KNOWNBUG).
9. Created the regression test invar_assigns_opt (CORE).
10. Created the regression test invar_assigns_empty (CORE).

This adds parser support for assigns contracts on loops.

feliperodri commented on regression/contracts/invar_assigns_alias_analysis/test.desc:
```diff
@@ -0,0 +1,14 @@
+^VERIFICATION FAILED$
+--
+--
+This test shows that the current implementation only supports
+parsing and still uses alias analysis. Fixing this is a part 
```
What do you mean by "current implementation only supports parsing and still uses alias analysis"? Could you elaborate?

aalok-thakkar replied:
This PR only adds parsing to it, and assertion statements are yet not generated. Therefore, this proof does not go through.

SaswatPadhi replied:
What Aalok is saying is that, he added this test just to check the parsing support (and address Martin's comment regarding usage of assigns clause). But it's not considered during instrumentation, so this test which he obtained from my bug report, still fails as before.

feliperodri commented on regression/contracts/invar_assigns_empty/test.desc:
```diff
@@ -0,0 +1,12 @@
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test checks that empty assigns clause is supported.
```
```
This test checks that empty assigns clause is supported in loop contracts.
```

feliperodri reviewed (approved):
LGTM.

feliperodri commented:
@martin-cs could you give another review on this PR?

martin-cs commented on regression/contracts/invar_assigns_alias_analysis/test.desc:
```diff
@@ -0,0 +1,14 @@
+KNOWNBUG
```
If it is marked as a known bug then the test will not be run.  Is it possible to either have just a test for the parsing or test for the "incorrect" results until it is implemented.

feliperodri replied:
@SaswatPadhi @aalok-thakkar

SaswatPadhi replied:
There is one such test already, to check just the parsing. Please see: regression/contracts/invar_assigns_opt/main.c.

martin-cs replied:
Thanks.

martin-cs reviewed (approved):
Sure @feliperodri .  There are now tests so no longer blocking.
